### PR TITLE
Map speedup

### PR DESCRIPTION
**Pull request by kopyl**
Speed up 2nd mapping in `examples/text_to_image/train_text_to_image_sdxl.py` (computing VAE).

Testing on 833 samples of this dataset: lambdalabs/pokemon-blip-captions

Mine: 1m 48s
Current implementation: 2m 25s

### CODE REVIEW

**Comment by kopyl** [comment]
@sayakpaul could you please merge i?

**Comment by sayakpaul** [comment]
Can you explain what exactly are you doing here?

**Comment by kopyl** [comment]
@sayakpaul sure. Instead of mapping over the already mapped dataset once, i assign it to another variable and then merge those datasets together.

Speeds things up a lot.

**Comment by sayakpaul** [comment]
@lhoestq could you comment on this change?

**Review by lhoestq — COMMENTED** [review]
I don't see how it can be faster since there are still the two same `map()` calls, can you elaborate ?

Also not sure why `with_transform` is called twice (once in your code addition, and once line 866)

**Comment by kopyl** [comment]
@lhoestq have no idea how it can be faster, but it is, you can run a test training on a Pokémon datasets and see that it's faster.

I don't know how to remove the second .with_transform (or first). If I do, it breaks the entire training due to the absence of required columns. Maybe you can help me do that? Also I can't understand why it's with_transform instead of just a plain map.

**Comment by sayakpaul** [comment]
@lhoestq a gentle ping.

**Comment by lhoestq** [comment]
I think it's faster because the second map call has to read+write the previously computed embeddings even though they're already saved, good find !

I would not add `train_dataset.cleanup_cache_files()` though, because it would not allow to reload the cache properly if you run the code a second time

**Comment by kopyl** [comment]
@lhoestq done. Please merge ❤️

**Comment by kopyl** [comment]
@lhoestq i have more very cool things i could add to the training script like saving the pre-computed dataset so it does not compute it when the same script is run again with different params. Just need to find some time to do it cause i also need to do some testing as well...

**Comment by kopyl** [comment]
@lhoestq merged the current main branch into this one

File: examples/text_to_image/train_text_to_image_sdxl.py [hunk 1]
```diff
@@ -35,7 +35,7 @@
 from accelerate import Accelerator
 from accelerate.logging import get_logger
 from accelerate.utils import ProjectConfiguration, set_seed
-from datasets import load_dataset
+from datasets import load_dataset, concatenate_datasets
```
Comment (lhoestq) [inline examples/text_to_image/train_text_to_image_sdxl.py:38]: fix import order
```suggestion
from datasets import concatenate_datasets, load_dataset
```

**Review by lhoestq — APPROVED** [review]
LGTM ! I'll let you merge @sayakpaul if it's good for you

PS: I also opened https://github.com/huggingface/diffusers/pull/7171 to fix an issue with the fingerprint of train_dataset_with_vae

File: examples/text_to_image/train_text_to_image_sdxl.py [hunk 2]
```diff
@@ -896,13 +896,19 @@ def preprocess_train(examples):
         # details: https://github.com/huggingface/diffusers/pull/4038#discussion_r1266078401
         new_fingerprint = Hasher.hash(args)
         new_fingerprint_for_vae = Hasher.hash("vae")
-        train_dataset = train_dataset.map(compute_embeddings_fn, batched=True, new_fingerprint=new_fingerprint)
-        train_dataset = train_dataset.map(
+        train_dataset_with_embeddings = train_dataset.map(
+            compute_embeddings_fn, batched=True, new_fingerprint=new_fingerprint
+        )
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_sdxl.py:901]: (nit) This can be a single line.

**Review by sayakpaul — APPROVED** [review]
Looking great! Just one small change.

**Comment by sayakpaul** [comment]
Let's also resolve the merge conflicts.

**Comment by kopyl** [comment]
@sayakpaul 

https://github.com/huggingface/diffusers/pull/6745#discussion_r1508821808

done

**Comment by sayakpaul** [comment]
There's a merge conflict that we need to resolve before we can ship this.

**Comment by kopyl** [comment]
@sayakpaul 

https://github.com/huggingface/diffusers/pull/6745#issuecomment-1973017410

I don't see any merge conflicts. So could you please help me find them so I can try to do that? :)

**Comment by kopyl** [comment]
> There's a merge conflict that we need to resolve before we can ship this.

@sayakpaul of course. done. please check this

**Comment by sayakpaul** [comment]
Thanks! Ping me once the CI run is complete :)

**Comment by sayakpaul** [comment]
But now we have a code quality problem :/

**Comment by kopyl** [comment]
> But now we have a code quality problem :/ 

What problem exactly? My change did not introduce any bugs.

**Comment by kopyl** [comment]
@sayakpaul i see the logs. But there is no info on what's exactly wrong.

Could you please tag the person who is responsible for the tests?

**Comment by sayakpaul** [comment]
You need to run `make style && make quality` to ensure the code is properly formatted.

**Comment by kopyl** [comment]
@sayakpaul thanks. Now all the checks seem to be completed

**Comment by sayakpaul** [comment]
Thanks a lot for your hard work.

**Comment by kopyl** [comment]
@sayakpaul u too, it was a pleasure ❤️. Be ready to review more cool stuff from me in the near future.

**Pull request merged**